ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -300,13 +304,21 @@ void fillChunks() {
           if (channelReadBuffer != null) {
             chunkToFill = getChunkToFill();
             if (chunkToFill == null) {
-              // no chunks are free to be filled yet.
+              // channel has data, but no chunks are free to be filled yet.
+              maybeStartTrackingWaitTime();
               break;
-            }
-            bytesFilledSoFar += chunkToFill.fillFrom(channelReadBuffer);
-            if (!channelReadBuffer.hasRemaining()) {
-              chunkFillerChannel.resolveOldestChunk(null);
-              channelReadBuffer = null;
+            } else {
+              // channel has data, and there is a chunk that can be filled.
+              maybeUpdateTrackedWaitTime();
+              bytesFilledSoFar += chunkToFill.fillFrom(channelReadBuffer);
+              if (chunkToFill.isReady()) {
+                routerMetrics.waitTimeForFreeChunkAvailabilityMs.update(waitTimeForNextChunkAvailabilityMs);
+                resetWaitTimeTracking();
```

Comment:
`maybeUpdateTrackedWaitTime()` is called every time we are about to fill data to a chunk - it is not necessarily the point where a free chunk becomes available - as the chunk in question could have been partially filled and left off previously (in BUILDING state) because the channel did not have data.

Also, we want to ensure that if two chunks became available, then the wait time for the second chunk is reported as 0.